src/unix/unix.c

```diff
@@ -412,66 +395,23 @@ void process_get_cwd(process p, filesystem *cwd_fs, inode *cwd)
     process_unlock(p);
 }
 
-void thread_enter_user(thread in)
-{
-    thread_resume(in);
-    in->sysctx = false;
-}
-
-void thread_enter_system(thread t)
-{
-    if (!t->sysctx) {
-        timestamp here = now(CLOCK_ID_MONOTONIC_RAW);
-        timestamp diff = here - t->start_time;
-        t->utime += diff;
-        t->start_time = here;
-        t->sysctx = true;
-        set_current_thread(&t->thrd);
-    }
-}
-
-void thread_pause(thread t)
-{
-    if (get_current_thread() != &t->thrd)
-        return;
-    timestamp diff = now(CLOCK_ID_MONOTONIC_RAW) - t->start_time;
-    if (t->sysctx) {
-        t->stime += diff;
-    }
-    else {
-        t->utime += diff;
-    }
-    context f = thread_frame(t);
-    thread_frame_save_fpsimd(f);
-    thread_frame_save_tls(f);
-    set_current_thread(0);
-}
-
-void thread_resume(thread t)
-{
-    nanos_thread old = get_current_thread();
-    if (old && old != &t->thrd)
-        apply(old->pause);
-    count_syscall_resume(t);
-    if (get_current_thread() == &t->thrd)
-        return;
-    t->start_time = now(CLOCK_ID_MONOTONIC_RAW);
-    set_current_thread(&t->thrd);
-}
-
 static timestamp utime_updated(thread t)
 {
+    thread_lock(t);
     timestamp ts = t->utime;
-    if (!t->sysctx)
+    if (t->start_time != 0)
         ts += now(CLOCK_ID_MONOTONIC_RAW) - t->start_time;
+    thread_unlock(t);
     return ts;
 }
 
 static timestamp stime_updated(thread t)
 {
+    thread_lock(t);
     timestamp ts = t->stime;
-    if (t->sysctx)
-        ts += now(CLOCK_ID_MONOTONIC_RAW) - t->start_time;
+    if (t->syscall && t->syscall->start_time)
+        ts += now(CLOCK_ID_MONOTONIC_RAW) - t->syscall->start_time;
```

Comment:
`t->syscall` can be modified without the thread lock, e.g. it's set to 0 in syscall_finish(), so there could be a null pointer access here. In fact, running the apache httpd web server with 2 CPUs I can replicate a condition where this happens.

Reply:
Rather than requiring the thread lock to access t->syscall, I removed utime_updated() and stime_updated() (both of which are unsafe, as the start_time fields are also not protected by the thread or other lock) and have the calling routines simply read t->utime and t->stime directly.

The downside is that the times will not be current to now(), but the assumption here is that the granularity given by times being accumulated on each context switch will give sufficient resolution and accuracy for the dependent syscalls (times, getrusage, and clock_gettime).